metadata-ingestion/src/datahub/ingestion/source/bigquery/bigquery_audit.py

```diff
@@ -0,0 +1,573 @@
+    project_id: str
+    dataset: str
+    table: str
+
+    _BIGQUERY_DEFAULT_SHARDED_TABLE_REGEX: str = "((.+)[_$])?(\\d{4,10})$"
+
+    def _get_table_and_shard(self, table_name: str) -> Tuple[str, Optional[str]]:
+        match = re.search(
+            BigqueryTableIdentifier._BIGQUERY_DEFAULT_SHARDED_TABLE_REGEX,
+            table_name,
+            re.IGNORECASE,
+        )
+        if match:
+            table_name = match.group(2)
+            shard = match.group(3)
+            return table_name, shard
+        return table_name, None
+
+    @classmethod
+    def from_string_name(cls, table: str) -> "BigqueryTableIdentifier":
+        parts = table.split(".")
+        return cls(parts[0], parts[1], parts[2])
+
+    def raw_table(self):
+        return f"{self.project_id}.{self.dataset}.{self.table}"
+
+    @staticmethod
+    def _remove_suffix(input_string, suffix):
+        if suffix and input_string.endswith(suffix):
+            return input_string[: -len(suffix)]
+        return input_string
+
+    def get_table(self) -> str:
+        shortened_table_name = self.table
+        # if table name ends in _* or * then we strip it as that represents a query on a sharded table
+        shortened_table_name = self._remove_suffix(shortened_table_name, "_*")
+        shortened_table_name = self._remove_suffix(shortened_table_name, "*")
+
+        table_name, _ = self._get_table_and_shard(shortened_table_name)
+        if not table_name:
+            table_name = self.dataset
+
+        matches = SNAPSHOT_TABLE_REGEX.match(table_name)
+        if matches:
+            table_name = matches.group(1)
+            logger.debug(f"Found table snapshot. Using {table_name} as the table name.")
+
+        # Handle exceptions
+        invalid_chars_in_table_name: List[str] = [
+            c for c in {"$", "@"} if c in table_name
+        ]
+        if invalid_chars_in_table_name:
+            raise ValueError(
+                f"Cannot handle {self} - poorly formatted table name, contains {invalid_chars_in_table_name}"
+            )
+
+        return f"{self.project_id}.{self.dataset}.{table_name}"
+
+    def __str__(self) -> str:
+        return self.get_table()
+
+
+@dataclass(frozen=True, order=True)
+class BigQueryTableRef:
+    table_identifier: BigqueryTableIdentifier
+
+    @classmethod
+    def from_bigquery_table(cls, table: BigqueryTableIdentifier) -> "BigQueryTableRef":
+        return cls(
+            BigqueryTableIdentifier(table.project_id, table.dataset, table.table)
+        )
+
+    @classmethod
+    def from_spec_obj(cls, spec: dict) -> "BigQueryTableRef":
+        return cls(
+            BigqueryTableIdentifier(
+                spec["projectId"], spec["datasetId"], spec["tableId"]
+            )
+        )
+
+    @classmethod
+    def from_string_name(cls, ref: str) -> "BigQueryTableRef":
+        parts = ref.split("/")
+        if parts[0] != "projects" or parts[2] != "datasets" or parts[4] != "tables":
```

Comment:
`or parts[2] != "datasets" or parts[4] != "tables"` can lead to IndexError if len(parts) is not checked before accessing elements at indices 2 & 4.